// nhentai/lib/doujin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Doujin = void 0;
const image_1 = require("./image");
const tag_1 = require("./tag");
const constants_1 = require("./constants");
class Doujin {
    constructor(raw) {
        this.raw = raw;
        this.id = raw.id;
        this.mediaId = Number(raw.media_id);
        this.titles = raw.title;
        this.scanlator = raw.scanlator.length > 0 ? raw.scanlator : null;
        this.uploadDate = new Date(raw.upload_date * 1000);
        this.uploadTimestamp = raw.upload_date;
        this.length = raw.num_pages;
        this.favorites = raw.num_favorites;
        this.url = `${constants_1.HOST_URL}/g/${raw.id}`;
        this.pages = raw.images.pages.map((image, index) => new image_1.Image(image, index + 1, this));
        this.cover = new image_1.Image(raw.images.cover, 'cover', this);
        this.thumbnail = new image_1.Image(raw.images.thumbnail, 'thumb', this);
        this.tags = new tag_1.TagManager(raw.tags);
    }
    hasTagByName(name) {
        return !!this.tags.all.find(tag => tag.name === name);
    }
    hasTagByID(ID) {
        return !!this.tags.all.find(tag => tag.id === ID);
    }
}
exports.Doujin = Doujin;
