// nhentai/lib/tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagManager = exports.Tag = void 0;
const constants_1 = require("./constants");
class Tag {
    constructor(raw) {
        this.id = raw.id;
        this.type = raw.type;
        this.name = raw.name;
        this.url = constants_1.HOST_URL + raw.url;
        this.count = raw.count;
    }
}
exports.Tag = Tag;
class TagManager {
    constructor(raw) {
        this.getById = (id) => this.all.find(tag => tag.id === id);
        this.getByType = (type) => this.all.filter(tag => tag.type === type);
        this.all = raw.map(tag => new Tag(tag));
    }
    get tags() {
        return this.getByType('tag');
    }
    get groups() {
        return this.getByType('group');
    }
    get languages() {
        return this.getByType('language');
    }
    get artists() {
        return this.getByType('artist');
    }
    get characters() {
        return this.getByType('character');
    }
    get parodies() {
        return this.getByType('parody');
    }
    get categories() {
        return this.getByType('category');
    }
}
exports.TagManager = TagManager;
